### PR TITLE
Oak 11113 - MissingLastRevSeekerTest fix flaky test

### PR DESCRIPTION
**Pull request by ionutzpi**
MissingLastRevSeekerTest fix flaky test getNonSplitDocs

### CODE REVIEW

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java [hunk 1]
```diff
@@ -238,6 +238,11 @@ private void getBrokenSurrogateAndInitializeDifferentStores(DocumentStoreFixture
             if (nodeStore != null) {
                 nodeStore.dispose();
             }
+            try {
```
Comment (reschke) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java]: In theory, disposing the node store should take care of this. We should find out why it doesn't.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java [hunk 2]
```diff
@@ -249,7 +251,7 @@ public void dispose() {
         }
         dns.runBackgroundOperations();
         //seeker should return only non split documents
-        int docs = Iterables.size(seeker.getCandidates(0));
+        int docs = Iterables.size(seeker.getCandidates(Objects.requireNonNull(dns.getRoot().getLastRevision().getRevision(dns.getClusterId())).getTimestamp()));
```
Comment (reschke) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Is this related to the flakyness?

Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Yes. I removed 0 and replaced it with proper revision version timestamp.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java (hunk 1 above)
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java]: It is just a precocious adding. No truly impact.

Comment (reschke) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java]: Ok, then please remove that change, because a) not needed, b) not related, and c) potentially confusing.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 2 above)
Comment (reschke) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Ok.

a) Could you put a short description of what was wrong into the ticket?

b) I would recommend that @egli, @rishabhdaim or @Joscorbe confirm that his is the correct change.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java (hunk 1 above)
Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java]: Done.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java (hunk 2 above)
Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: `0` was used to fetch all the documents. Using the timestamp of `lastRevision` would result in skipping some documents.

If there is no proper reason, I will revert the change.

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: I was able to reproduce this test failure on my local. The root cause the is presence of document with id `1:/test` in the `RDB-H2`.

Once I did `mvn clean install` this problem went away.
IMO, it is because of some earlier test(s) didn't cleanup properly rather than test.

Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Yes. That is the reason that I modified the timestamp to return only the documents needed, not all documents.
@reschke we can have inconsistencies retrieving all docs from RDB, but retrieving only the ones that this test added should solve the issue and future issues that could appear regarding this test. What do you think?

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: I think I found the Root cause.

These tests are failing due to the `CompressedDocumentPropertyStateTest` test not cleaning up properly.
On my local, it is causing either Mongo or RDB-H2 to fail.

`CompressedDocumentPropertyStateTest` creates the problematic node `1:/test` causing test failures.

`CompressedDocumentPropertyStateTest` has been added recently which explains the recent failures of `MissingLastRevSeekerTest`.

@ionutzpi could you please look into this?
cc @Joscorbe @reschke

Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Yes. I added CompressedDocumentPropertyStateTest changes to solve this issue. Julian told that documentfixturestore should close when closing nodestore(seems that this is not true).

Comment (rishabhdaim) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: Calling `fixture.dispose();` after disposing of `nodeStore` [1] would fix this problem for `Mongo`.

[1] https://github.com/apache/jackrabbit-oak/blob/trunk/oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyStateTest.java#L240

Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: It is not removing for RDB, I dont know the reason yet.

Comment (ionutzpi) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/MissingLastRevSeekerTest.java]: @rishabhdaim if I add removeMe.add("1:/test"); after assert it will fix the issue. Do you agree with this change?

**Comment by rishabhdaim** [comment]
Created PR https://github.com/apache/jackrabbit-oak/pull/1749 with proposed changes.

**Comment by reschke** [comment]
@rishabhdaim - thanks for the analysis. This also explains why the test hasn't been flaky for a long time.

Note that since a few weeks, AbstractDocumentStoreTest will check the NODES table before/after. So if a test fails like that, a check of the unit test log might be interesting.

@ionutzpi - trust me; the RDB based tests assume that the test nodes are cleaned up (unless the test has been designed to purge tables). This is different from MongoDB because back then, I didn't want tests to nuke tables they did not create. This may have been the wrong approach, and we can discuss to change this. I'm just not sure it's worth the effort.

**Comment by reschke** [comment]
Once it's clear which test is flaky, please adjust the issue title and make sure that the PR commit message matches.

**Comment by rishabhdaim** [comment]
@reschke I have created PR to fix the flacky tests here https://github.com/apache/jackrabbit-oak/pull/1749.
Could you please review the same. Also the build passes after the proposed fix.
